src/librustc_interface/util.rs

```diff
@@ -109,6 +108,9 @@ pub fn create_session(
     let codegen_backend = get_codegen_backend(&sess);
 
     rustc_lint::register_builtins(&mut sess.lint_store.borrow_mut(), Some(&sess));
+    if sess.unstable_options() {
```

Comment:
Why was it guarded by `unstable_options()`?

Reply:
Because every lint, that is registered will be run, which is bad performance wise (for normal users) and can be easily prevented. We decided to not introduce a new `-Z` option and just use the `unstable_options` flag. https://github.com/rust-lang/rust/pull/58701#discussion_r261378083

Reply:
These lints are useless outside of the compiler or clippy. Exposing them to stable would mean we'd need to support the name (or the lint existing) forever, even if the lint might disappear if the compiler datastructures change beyond the point where the lint is needed

Reply:
Oh, GitHub collapsed that comment.

I was wondering if there is better solution for Clippy and possibly other projects.

Reply:
Why not gating on whether the crate is compiled to be unstable with the feature `rustc_private`?